src/common/dav/tx.c

```diff
@@ -1491,11 +1493,22 @@ dav_tx_off2ptr(uint64_t off)
 uint64_t
 dav_reserve(dav_obj_t *pop, struct dav_action *act, size_t size, uint64_t type_num)
 {
+	int tx_inprogress = 0;
+	int rc;
+
 	DAV_DBG("pop %p act %p size %zu type_num %llx",
 		pop, act, size,
 		(unsigned long long)type_num);
 
+	if (get_tx()->stage != DAV_TX_STAGE_NONE)
+		tx_inprogress = 1;
+
 	DAV_API_START();
+	if (!tx_inprogress) {
+		rc = lw_tx_begin(pop);
+		if (rc)
+			return rc;
+	}
 	if (pop->do_utx == NULL && dav_umem_wtx_new(pop) == NULL)
```

Comment:
I will correct the check also the return value.

On the comment on UMEM_COMMIT_DEFER, the existing code under PMDK commits the memcpy() no mater whether the transaction under which it is called succeeds or aborts(). If the gc_bin_add_item() works fine if memcpy is not committed when the transaction aborts, I can replace existing code with memcpy. Else, I can add an assert under dav_memcpy_persist_relaxed() stating that it should be called only within a transaction. This will ensure that the issue similar to dav_reserve will not happen.

Reply:
As for the gc_bin_add_item(), I think the original intention of calling pmemobj_memcpy_persist() is just to avoid making snapshot (see the comment in gc_bin_add_item()), this actually can be achieved by "umem_tx_xadd_ptr(NO_SNAPSHOT) + regular memcpy()". @jolivier23 , @gnailzenh , is that correct?

Reply:
I have replaced atomic copy in gc_bin_add_item() with umem_tx_xadd_ptr(NO_SNAPSHOT) + regular memcpy(). I have retained the flag UMEM_RESERVED_MEM which is used in bio_memcpy(). The comment there states that with pmdk atomic copy there is better than allowing it to be drained in the subsequent publish + tx_commit.